room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/TypeConverterFactoryTest.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2018 The Android Open Source Project
```

Comment:
can we expand this test a little bit more to have cases where a type converter is declared in a lower layer, like a dao and entity.

for instance, while trying to understand why we need the add factory callback in Context, i removed it and all of these tests still pass since it only tests for factories declared in the database.

Reply:
repro for that case:
https://android-review.googlesource.com/c/platform/frameworks/support/+/1419740

Reply:
I'll look into that, thanks.

Reply:
Yeah seems like keeping the add factory callback in Context may not be needed. It's only needed to be passed to `fork` method because this method calls `findConverters` logic.

I'll add more tests.

Reply:
i have a prototype for the DAO based suggestion, will upload it soon.

Reply:
here is the *dirty* prototype for removing it from context fully.
https://android-review.googlesource.com/c/platform/frameworks/support/+/1420426

Basically what i do is:
* add a function to codegen scope that is special for converter factories (ugly)
* in dao, generate a method that will return all of the factory `Class`es (not strings)
* in db, generate a method that will return a `Map<String, List<Class<? : TypeConverterFactory>>` that is a map of DAO name to the list of factories it requires.
* in database constructor, get that list and verify against provided items.

I also made another change to not use the cannonical names as keys, instead, just use assignability.
I'm not sure why we preferred strings so this change might be wrong but i thought it would be handy for cases where a sub-type of the factory will be provided.

Lmk what you think. This is just a prototype, quite a dirty implementation but passes tests.

Reply:
cc @danysantiago

Reply:
Seema like there is not much left to be done here! 😛 Nice work

Reply:
it needs some cleanup but yea i'm not unhappy about it. I'm not sure if it is better though, lets see what @danysantiago thinks and then if we'll go with that, i can send a PR to your fork.

You can probably already apply it to here via:
```
git fetch https://android.googlesource.com/platform/frameworks/support refs/changes/26/1420426/2 && git cherry-pick FETCH_HEAD
```
Not sure if that command will require authentication, if not, you just need to run it in your local checkout in theory.

Though I should first clean it up.

Reply:
These are valid points, I like some of Yigit's approach, the DAO informing the factory it expects and then the Database gathering those for validation. However, I'm not convinced about using the `ClassWritter` and adding type converter factory specific code. I talked with @yigit offline (as in, online via a video call) and he is going to cleanup the usage of `ClassWitter` a bit more by having a more general solution for metadata, which in this case would be factories used. Ultimately, @mzgreen, you should able to cherry-pick his patch, adjust it if necessary and then get this merged.

Reply:
here is that change:
https://android-review.googlesource.com/c/platform/frameworks/support/+/1420426/2..4

lmk what you think and i'll clean it up for @mzgreen to cherry-pick.

Reply:
ok that implementation is ready for review as well:
https://android-review.googlesource.com/c/platform/frameworks/support/+/1420426

@danysantiago @mzgreen lmk what you prefer on whether we want to review that separately since this one is already reviewed.

The changes in that CL are:
* use class as key, allow subclasses
* collect factories in DAOs
* remove get converter factory map from RoomDatabase but instead provide a getter. This will allow not exposing the map so that it is easier to change in the future if we need to.